zokrates_embed/src/ark.rs

```diff
@@ -0,0 +1,344 @@
+        || Ok(proof),
+        AllocationMode::Witness,
+    )
+    .unwrap();
+
+    let vk = <VerifierGadget as SNARKGadget<
+        <BLS12PairingEngine as PairingEngine>::Fr,
+        <BLS12PairingEngine as PairingEngine>::Fq,
+        GM17Snark,
+    >>::new_verification_key_unchecked(
+        ns!(cs, "alloc_vk"), || Ok(vk), AllocationMode::Witness
+    )
+    .unwrap();
+
+    let res = <VerifierGadget as SNARKGadget<
+        <BLS12PairingEngine as PairingEngine>::Fr,
+        <BLS12PairingEngine as PairingEngine>::Fq,
+        GM17Snark,
+    >>::verify(&vk, &inputs, &proof)
+    .unwrap();
+
+    cs.finalize();
+
+    let num_instance_variables = cs.num_instance_variables();
+    let input_indices = fp_vars
+        .iter()
+        .map(|f| var_to_index(&f, 0))
+        .collect::<Vec<usize>>();
+
+    let proof_indices: Vec<usize> = vec![
+        var_to_index(&proof.a.x, num_instance_variables),
+        var_to_index(&proof.a.y, num_instance_variables),
+        var_to_index(&proof.b.x.c0, num_instance_variables),
+        var_to_index(&proof.b.x.c1, num_instance_variables),
+        var_to_index(&proof.b.y.c0, num_instance_variables),
+        var_to_index(&proof.b.y.c1, num_instance_variables),
+        var_to_index(&proof.c.x, num_instance_variables),
+        var_to_index(&proof.c.y, num_instance_variables),
+    ];
+
+    let mut vk_indices: Vec<usize> = vec![
+        var_to_index(&vk.h_g2.x.c0, num_instance_variables),
+        var_to_index(&vk.h_g2.x.c1, num_instance_variables),
+        var_to_index(&vk.h_g2.y.c0, num_instance_variables),
+        var_to_index(&vk.h_g2.y.c1, num_instance_variables),
+        var_to_index(&vk.g_alpha_g1.x, num_instance_variables),
+        var_to_index(&vk.g_alpha_g1.y, num_instance_variables),
+        var_to_index(&vk.h_beta_g2.x.c0, num_instance_variables),
+        var_to_index(&vk.h_beta_g2.x.c1, num_instance_variables),
+        var_to_index(&vk.h_beta_g2.y.c0, num_instance_variables),
+        var_to_index(&vk.h_beta_g2.y.c1, num_instance_variables),
+        var_to_index(&vk.g_gamma_g1.x, num_instance_variables),
+        var_to_index(&vk.g_gamma_g1.y, num_instance_variables),
+        var_to_index(&vk.h_gamma_g2.x.c0, num_instance_variables),
+        var_to_index(&vk.h_gamma_g2.x.c1, num_instance_variables),
+        var_to_index(&vk.h_gamma_g2.y.c0, num_instance_variables),
+        var_to_index(&vk.h_gamma_g2.y.c1, num_instance_variables),
+    ];
+
+    vk_indices.extend(
+        vk.query
+            .iter()
+            .map(|q| {
+                vec![
+                    var_to_index(&q.x, num_instance_variables),
+                    var_to_index(&q.y, num_instance_variables),
+                ]
+            })
+            .flatten(),
+    );
+
+    let out_index = match &res {
+        Boolean::Is(x) => x
+            .variable()
+            .get_index_unchecked(num_instance_variables)
+            .unwrap(),
+        Boolean::Not(x) => x
```

Comment:
I find it surprising that we would return `x` for both `x` and `!x`? Are both of these match arms reachable? Or are we missing a test somewhere?

Reply:
actually `Not` variant seems to be unreachable, nice catch